### PR TITLE
[ASUSGAB] User provisioning API

### PR DESCRIPTION
**Pull request by AndreyLykhoman**
**Description:** 
1. Add User provisioning API with two endpoints:
 - create a user without password
 - bulk enroll users
2. Addd phone field to user profile.

**Youtrack:** https://youtrack.raccoongang.com/issue/ASUSGAB-3

**Settings**: 
Add to lms.auth.json 'EDX_API_KEY'. It's key for check valid token.
The following token should be available in the python code:
'HTTP_X_EDX_API_KEY,

X-Edx-Api-Key HTTP header is present in the request

### CODE REVIEW

File: lms/djangoapps/sga_api/views.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+            user.is_active = True
+            user.save()
+            user_profile, _ = UserProfile.objects.get_or_create(user=user)
+            user_profile.phone = request.data.get('phone', '')
+            user_profile.save()
+            # dict to setup language
+            update = {
+                u'pref-lang': language,
+            }
+            #setup language for user
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Space after sharp is required.

File: lms/djangoapps/sga_api/views.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+            data['password'] = uuid4().hex
+            data['country'] = country
+            user = create_account_with_params(request, data)
+            user.is_active = True
+            user.save()
+            user_profile, _ = UserProfile.objects.get_or_create(user=user)
+            user_profile.phone = request.data.get('phone', '')
+            user_profile.save()
+            # dict to setup language
+            update = {
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Why do you use multi-string dict for the small object?

File: lms/djangoapps/sga_api/views.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+                u'pref-lang': language,
+            }
+            #setup language for user
+            update_user_preferences(user, update, username)
+            self.send_activation_email(request)
+        except ValidationError:
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Is it possible to get the validation error different from the email?

File: lms/djangoapps/sga_api/views.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+            return Response(data={"error_message": "Wrong email format"}, status=status.HTTP_400_BAD_REQUEST)
+
+        return Response(data={'user_id': user.id, 'username': username}, status=status.HTTP_200_OK)
+
+    @staticmethod
+    def send_activation_email(request):
+        form = PasswordResetFormNoActive(request.data)
+        if form.is_valid():
+            form.save(use_https=request.is_secure(),
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: The OpenEdx style guide is required.

File: lms/djangoapps/sga_api/views.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+                          'email_from_address', settings.DEFAULT_FROM_EMAIL),
+                      request=request,
+                      subject_template_name='sga_api/set_password_subject.txt',
+                      email_template_name='sga_api/set_password_email.html'
+                      )
+
+
+
+class BulkEnrollView(APIView, ApiKeyPermissionMixIn):
+    authentication_classes = (OAuth2AuthenticationAllowInactiveUser,)
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: ditto L31

File: lms/djangoapps/sga_api/views.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+    permission_classes = ApiKeyHeaderPermission,
+
+    def post(self, request):
+        """
+        Enrolling user on the course with the specified mod.
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: ditto L35

File: lms/djangoapps/sga_api/views.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        users = data.get('users')
+        courses = data.get('courses')
+        action = data.get('action', 'enroll')
+        email_students = data.get('email_students', False)
+        auto_enroll = data.get('auto_enroll', False)
+
+        if action not in ['enroll', 'unenroll']:
+            error_msg = "action must be enroll or unenroll"
+        elif not isinstance(courses, (list, tuple)):
+            error_msg = "courses_id must be a list of courses_id "
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: There is no parameter `courses_id` suppose it is typo

File: lms/djangoapps/sga_api/views.py [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+        elif not isinstance(users, (list, tuple)):
+            error_msg =  "users must be a list of identifiers(username or email)"
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: If users = ('username', 'email') would it be enough for this check?

File: lms/djangoapps/sga_api/views.py [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+
+        if error_msg:
+            return Response(
+                data={"error_message": error_msg},
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            result = self.__enroll_unenrol_users_for_courses(
+                courses, users, request,
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Please do not split the line if it is not needed.

File: lms/djangoapps/sga_api/views.py [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+            return Response(
+                data={
+                    'action': action,
+                    'courses': result,
+                    'email_students':email_students,
+                    'auto_enroll': auto_enroll,
+                },
+                status=status.HTTP_200_OK
+            )
+        except  Exception as e:
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: too many spaces before Exception

File: lms/djangoapps/sga_api/views.py [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+            form.save(use_https=request.is_secure(),
+                      from_email=configuration_helpers.get_value(
+                          'email_from_address', settings.DEFAULT_FROM_EMAIL),
+                      request=request,
+                      subject_template_name='sga_api/set_password_subject.txt',
+                      email_template_name='sga_api/set_password_email.html'
+                      )
+
+
+
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Odd empty line

File: lms/djangoapps/sga_api/views.py [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+            error_msg = None
+
+        if error_msg:
+            return Response(
+                data={"error_message": error_msg},
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            result = self.__enroll_unenrol_users_for_courses(
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Two underscores starting the function's name is not a good choice, please use one.

File: lms/djangoapps/sga_api/views.py [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+        except  Exception as e:
+            log.error(e.message)
+            return Response(
+                data={"error_message": e.message},
+                status=status.HTTP_500_INTERNAL_SERVER_ERROR,
+            )
+
+    def __enroll_unenrol_users_for_courses(
+            self, courses_id, identifiers, request,
+            action='enroll', email_students=False, auto_enroll=False
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Please do not split lines if it is not required.

File: lms/djangoapps/sga_api/views.py [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+    ):
+        courses = {}
+        for course_id in courses_id:
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: That is a great naming!

File: lms/djangoapps/sga_api/views.py [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+            )
+        except  Exception as e:
+            log.error(e.message)
+            return Response(
+                data={"error_message": e.message},
+                status=status.HTTP_500_INTERNAL_SERVER_ERROR,
+            )
+
+    def __enroll_unenrol_users_for_courses(
+            self, courses_id, identifiers, request,
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Are you sure that `identifiers` is a good choice for the list of users?

File: lms/djangoapps/sga_api/views.py [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+            #setup language for user
+            update_user_preferences(user, update, username)
+            self.send_activation_email(request)
+        except ValidationError:
+            return Response(data={"error_message": "Wrong email format"}, status=status.HTTP_400_BAD_REQUEST)
+
+        return Response(data={'user_id': user.id, 'username': username}, status=status.HTTP_200_OK)
+
+    @staticmethod
+    def send_activation_email(request):
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Please add descriptions for all non-documented functions.

File: lms/djangoapps/sga_api/views.py [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+                try:
+                    user = get_student_from_identifier(identifier)
+                    email = user.email
+                    language = get_user_email_language(user)
+                except User.DoesNotExist as e:
+                    if auto_enroll and action == 'enroll':
+                        email = identifier
+                        language = None
+                    else:
+                        raise e
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Specifying exception is not needed if you what to raise the same one which was caught by the try-except block.

File: lms/djangoapps/sga_api/views.py [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+        for identifier in identifiers:
+            try:
+                course_id_obj = SlashSeparatedCourseKey.from_deprecated_string(course_id)
+                course = get_course_by_id(course_id_obj)
+                if email_students:
+                    email_params = get_email_params(course, auto_enroll, secure=request.is_secure())
+                else:
+                    email_params = None
+
+                try:
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Why do we need the new one try - except block?

File: lms/djangoapps/sga_api/views.py [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+                    user = get_student_from_identifier(identifier)
+                    email = user.email
+                    language = get_user_email_language(user)
+                except User.DoesNotExist as e:
+                    if auto_enroll and action == 'enroll':
+                        email = identifier
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: This is not clear why do we assign identifier as email?

File: lms/djangoapps/sga_api/views.py [hunk 20]
```diff
@@ -0,0 +1,276 @@
...
+                        language = None
+                    else:
+                        raise e
+
+                validate_email(email)
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: The line seems to be odd?

File: lms/djangoapps/sga_api/views.py [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+                if action == 'enroll':
+                    before, after, _ = enroll_email(
+                        course_id_obj, email, auto_enroll, email_students, email_params, language=language
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Why do you separate `language` parameter from others?

File: lms/djangoapps/sga_api/views.py [hunk 22]
```diff
@@ -0,0 +1,276 @@
...
+                    )
+                )
+            except User.DoesNotExist:
+                result_list.append(
+                    self._make_result_dict_for_identifier(
+                        identifier=identifier,
+                        error='User with {} identifier does not exist'.format(identifier)
+                    )
+                )
+            except Http404:
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: That is not a good decision to use views function it would better to use `modulestore().get_course()`

File: lms/djangoapps/sga_api/views.py [hunk 23]
```diff
@@ -0,0 +1,276 @@
...
+                    self._make_result_dict_for_identifier(
+                        identifier=identifier,
+                        error='Problem winth {}ing user. Error: {}'.format(action, exc)
+                    )
+                )
+                log.error(exc)
+        return result_list
+
+    def _make_result_dict_for_identifier(
+            self, identifier, after={}, before={}, error=''
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Do we need default values?

File: lms/djangoapps/sga_api/views.py [hunk 24]
```diff
@@ -0,0 +1,276 @@
...
+                result_list.append(
+                    self._make_result_dict_for_identifier(
+                        identifier=identifier,
+                        error='Problem winth {}ing user. Error: {}'.format(action, exc)
+                    )
+                )
+                log.error(exc)
+        return result_list
+
+    def _make_result_dict_for_identifier(
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: It is staticmethod

File: lms/djangoapps/sga_api/views.py [hunk 25]
```diff
@@ -0,0 +1,276 @@
...
+            self, identifier, after={}, before={}, error=''
+    ):
+        return {
+            'identifier': identifier,
+            'after': after,
+            'before': before,
+            'error': error,
+        }
+
+    def _make_result_dict_for_course(self, results_list, action='enroll', auto_enroll=False):
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: ditto

File: lms/templates/sga_api/set_password_email.html [hunk 26]
```diff
@@ -0,0 +1,13 @@
+{% load i18n %}{% load url from future %}{% autoescape off %}
+{% blocktrans %}You're receiving this e-mail because an account has been created for you on {{ site_name }}.{% endblocktrans %}
+
+{% trans "Please click on the following link to choose a new password and activate your account:" %}
+{% block reset_link %}
+{{ protocol }}://{{ site_name }}{% url 'student.views.password_reset_confirm_wrapper' uidb36=uid token=token %}
+{% endblock %}
+
+{% trans "Thanks for using our site!" %}
```
Comment (idegtiarov) [inline lms/templates/sga_api/set_password_email.html]: The `site` should be changed to the `site_name` -- parameter

File: openedx/core/djangoapps/user_api/urls.py [hunk 27]
```diff
@@ -9,6 +9,7 @@
 from .accounts.views import AccountDeactivationView, AccountViewSet
 from .preferences.views import PreferencesDetailView, PreferencesView
 from .verification_api.views import PhotoVerificationStatusView
+from lms.djangoapps.sga_api.views import CreateUserAccountWithoutPasswordView, BulkEnrollView
```
Comment (idegtiarov) [inline openedx/core/djangoapps/user_api/urls.py]: It is a wrong place for the package import. And also a wrong place for the new api endpoint urls file.
Move it to the `sga_api` dir.

**Comment by idegtiarov** [comment]
In general, I want to ask you: please start using linter before you open a PR as well as language (English) checker. Thank you in advance.

File: lms/djangoapps/sga_api/README.md [hunk 28]
```diff
@@ -0,0 +1,53 @@
+App has endpoints fo create user and bulk enroll users.
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/README.md]: I used your version.

File: lms/djangoapps/sga_api/README.md [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+
+Configuration instructions: Code need next additional settings:
+
+Add to lms.auth.json 'EDX_API_KEY'. It's key for check valid token.
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/README.md]: I changed this string. Now is: 
'''
'EDX_API_KEY' - It is token for checking authorized API calls. lms.auth.json must have this token.
'''
is it right?

File: common/djangoapps/student/models.py [hunk 30]
```diff
@@ -292,6 +292,8 @@ class Meta(object):
     bio = models.CharField(blank=True, null=True, max_length=3000, db_index=False)
     profile_image_uploaded_at = models.DateTimeField(null=True, blank=True)
 
+    phone = models.CharField(blank=True, max_length=255, db_index=True)
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply common/djangoapps/student/models.py]: I changed it. I use 50 char. Length of a phone with some separators is low than 50. Example: "+38 (097) 999 - 99 - 99". This phone number contains 23 char.

File: lms/djangoapps/sga_api/views.py [hunk 31]
```diff
@@ -0,0 +1,276 @@
+import logging
+from uuid import uuid4
+from django.http import Http404
+from django.conf import settings
+from django.core.exceptions import NON_FIELD_ERRORS, ValidationError
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: I deleted it.

File: lms/djangoapps/sga_api/views.py [hunk 32]
```diff
@@ -0,0 +1,276 @@
...
+from lms.djangoapps.instructor.enrollment import get_email_params, enroll_email, get_user_email_language, unenroll_email
+from lms.djangoapps.instructor.views.tools import get_student_from_identifier
+from courseware.courses import get_course_by_id
+
+log = logging.getLogger(__name__)
+
+
+class CreateUserAccountWithoutPasswordView(APIView):
+    authentication_classes = (OAuth2AuthenticationAllowInactiveUser,)
+    permission_classes = ApiKeyHeaderPermission,
```
[the remark this answers is not in the document]
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: Done

**Comment by AndreyLykhoman** [comment]
I looked and fixed the code according to the remarks.

File: lms/djangoapps/sga_api/views.py [hunk 33]
```diff
@@ -0,0 +1,293 @@
...
+        first_name = request.data.get('first_name', '')
+        last_name = request.data.get('last_name', '')
+        language = request.data.get('language')
+        if not username:
+            error_msg = "username is required parameter."
+        elif not email:
+            error_msg =  "email is required parameter."
+        elif not country:
+            error_msg = (
+                "Country is wrong or absent: {country}. For checking: Visit https://www.iso.org/obp . "
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: I would rephrase s/Country is wrong or absent/Country is incorrect or missed/

File: lms/djangoapps/sga_api/views.py [hunk 34]
```diff
@@ -0,0 +1,293 @@
...
+            error_msg = None
+
+        if error_msg:
+            return Response(
+                data={"error_message": error_msg},
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            validate_slug(username)
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Why do we need this separate validation? username is validated during create_account_with_params process.

File: lms/djangoapps/sga_api/views.py [hunk 35]
```diff
@@ -0,0 +1,293 @@
...
+        if check_account_exists(username=username, email=email):
+            return Response(data={"error_message": "User already exists"}, status=status.HTTP_409_CONFLICT)
+
+        try:
+            data['password'] = uuid4().hex
+            data['country'] = country
+            user = create_account_with_params(request, data)
+            user.is_active = True
+            user.save()
+            user_profile, _ = UserProfile.objects.get_or_create(user=user)
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: create is redundant UserProfile is created on the create_account_with_params step and can be gotten by `user.profile`

File: lms/djangoapps/sga_api/views.py [hunk 36]
```diff
@@ -0,0 +1,293 @@
...
+            user_profile.phone = request.data.get('phone', '')
+            user_profile.save()
+            # dict to setup language
+            update = {u'pref-lang': language}
+            # setup language for user
+            update_user_preferences(user, update, username)
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: username is also a redundant parameter in `update_user_preferences` function.

File: lms/djangoapps/sga_api/views.py [hunk 37]
```diff
@@ -0,0 +1,293 @@
...
+        users = data.get('users')
+        courses = data.get('courses')
+        action = data.get('action', 'enroll')
+        email_students = data.get('email_students', False)
+        auto_enroll = data.get('auto_enroll', False)
+
+        if action not in ['enroll', 'unenroll']:
+            error_msg = "action must be enroll or unenroll"
+        elif not isinstance(courses, (list, tuple)):
+            error_msg = "courses must be a list of courses id "
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: typo s/ list of courses id/ list of courses' ids/ as an option I would check whether courses is str and create a tuple of courses:
```
if isinstance(courses, str):
    courses = (courses,)
```

File: lms/djangoapps/sga_api/views.py [hunk 38]
```diff
@@ -0,0 +1,293 @@
...
+        elif not isinstance(users, (list, tuple)):
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: ditto

File: lms/djangoapps/sga_api/views.py [hunk 39]
```diff
@@ -0,0 +1,293 @@
...
+                 results, action, auto_enroll
+            )
+        return courses
+
+    def _enroll_unenrol_users_for_course(
+            self, course_id, list_of_users_email, request,
+            action='enroll', email_students=False, auto_enroll=False
+    ):
+        result_list = []
+        for users_email in list_of_users_email:
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: I would use singular value s/users_email/user_email/

File: lms/djangoapps/sga_api/views.py [hunk 40]
```diff
@@ -0,0 +1,293 @@
...
+            )
+
+    def _enroll_unenrol_users_for_courses(
+        self, list_of_courses_id, list_of_users_email,
+        request, action='enroll', email_students=False, auto_enroll=False
+    ):
+        courses = {}
+        for course_id in list_of_courses_id:
+            results = self._enroll_unenrol_users_for_course(
+                course_id, list_of_users_email, request,
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Two strings for function's parameters is not an option please do as following: 
```
results = self._enroll_unenrol_users_for_course(
                course_id, list_of_users_email, request,
)
```
or
```
results = self._enroll_unenrol_users_for_course(
    course_id, 
    list_of_users_email, 
    request,
)
```

File: lms/djangoapps/sga_api/views.py [hunk 41]
```diff
@@ -0,0 +1,293 @@
...
+                action, email_students, auto_enroll
+            )
+            courses[course_id] = self._make_result_dict_for_course(
+                 results, action, auto_enroll
+            )
+        return courses
+
+    def _enroll_unenrol_users_for_course(
+            self, course_id, list_of_users_email, request,
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: ditto

File: lms/djangoapps/sga_api/views.py [hunk 42]
```diff
@@ -0,0 +1,293 @@
...
+                else:
+                    result_list.append(
+                        self._make_result_dict_for_identifier(
+                            users_email=users_email,
+                            error='User with {} email does not exist'.format(users_email),
+                            after={}, before={}
+                        )
+                    )
+                    continue
+            try:
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: too many try - except blocks one would be fair enough.

File: lms/djangoapps/sga_api/views.py [hunk 43]
```diff
@@ -0,0 +1,293 @@
...
+    ):
+        result_list = []
+        for users_email in list_of_users_email:
+    
+            try:
+                user = get_student_from_identifier(users_email)
+                email = user.email
+                language = get_user_email_language(user)
+            except User.DoesNotExist:
+                if auto_enroll and action == 'enroll':
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: Could you please explain what for is this statement branch?

File: lms/djangoapps/sga_api/views.py [hunk 44]
```diff
@@ -0,0 +1,293 @@
...
+                    self._make_result_dict_for_identifier(
+                        users_email=users_email, after=after.to_dict(), before=before.to_dict(),
+                    )
+                )
+            except ValidationError:
+                result_list.append(
+                    self._make_result_dict_for_identifier(
+                        users_email=users_email,
+                        error='User with {} users_email has not valid email'.format(users_email),
+                        after={}, before={}
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: All things with the code duplication should be created as default parameters of the function.

File: lms/djangoapps/sga_api/views.py [hunk 45]
```diff
@@ -0,0 +1,293 @@
...
+
+                result_list.append(
+                    self._make_result_dict_for_identifier(
+                        users_email=users_email, after=after.to_dict(), before=before.to_dict(),
+                    )
+                )
+            except ValidationError:
+                result_list.append(
+                    self._make_result_dict_for_identifier(
+                        users_email=users_email,
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: users_email key is not required for not named parameter

File: lms/djangoapps/sga_api/views.py [hunk 46]
```diff
@@ -0,0 +1,293 @@
...
+                        error='User with {} users_email has not valid email'.format(users_email),
+                        after={}, before={}
+                    )
+                )
+            except InvalidKeyError:
+                result_list.append(
+                    self._make_result_dict_for_identifier(
+                        users_email=users_email,
+                        error='Wrong course_id: {}. Course does not exist'.format(course_id),
+                        after={}, before={}
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: ditto

File: lms/djangoapps/sga_api/views.py [hunk 47]
```diff
@@ -0,0 +1,293 @@
...
+                    )
+                )
+            except Exception as exc:
+                result_list.append(
+                    self._make_result_dict_for_identifier(
+                        users_email=users_email, error='Problem winth {}ing user. Error: {}'.format(action, exc),
+                        after={}, before={}
+                    )
+                )
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: ditto

File: lms/djangoapps/sga_api/views.py (hunk 34 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: My code at first check params and at next to create a user. If the method 'create_account_with_params' will change and will remove validation, my code will not crush. 
And my code doesn't do search user if username isn't valid.

File: lms/djangoapps/sga_api/views.py (hunk 33 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: I changed the message.

File: lms/djangoapps/sga_api/views.py (hunk 35 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: I changed them. Thank you for

File: lms/djangoapps/sga_api/views.py (hunk 36 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: I deleted the redundant parameter.

File: lms/djangoapps/sga_api/views.py (hunk 37 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: It's better. Thank you.

File: lms/djangoapps/sga_api/views.py (hunk 38 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: I changed them.

File: lms/djangoapps/sga_api/views.py (hunk 40 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: I did them by  this page form edx documentation:
https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#breaking-long-lines

I made a recommendation on the account settings editor and configured my editor according to the requirements on the same page.

File: lms/djangoapps/sga_api/views.py (hunk 41 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: Like L195

File: lms/djangoapps/sga_api/views.py (hunk 42 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: I use the first try-except block for check user. The user must be enabled if auto_enroll=false. But if auto_enroll=true user can absent. And in this case, the code can continue to work with email from the request.

File: lms/djangoapps/sga_api/views.py (hunk 43 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: if auto_enroll=true user can absent. And in this case, the code can continue to work with email from the request.

File: lms/djangoapps/sga_api/views.py (hunk 46 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: I removed them

File: lms/djangoapps/sga_api/views.py (hunk 47 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: I removed them

File: lms/djangoapps/sga_api/views.py (hunk 45 above)
Comment (AndreyLykhoman) [reply lms/djangoapps/sga_api/views.py]: I think if I skip some of the input parameters, then named parameters will help them escape errors.

File: lms/djangoapps/sga_api/views.py [hunk 48]
```diff
@@ -0,0 +1,276 @@
+import logging
```
[the remark this answers is not in the document]
Comment (idegtiarov) [reply lms/djangoapps/sga_api/views.py]: 👍

File: lms/djangoapps/sga_api/views.py [hunk 49]
```diff
@@ -0,0 +1,293 @@
...
+        Creates a user using mail, login and also name and surname.
+        Sets a random password and sends a user a message to change it.
+        """
+        data = request.data
+        data['honor_code'] = "True"
+        data['terms_of_service'] = "True"
+        email = request.data.get('email')
+        username = request.data.get('username')
+        # countries.by_name return country code or '' for else cases
+        country = countries.by_name(request.data.get('country', ''))
```
[the remark this answers is not in the document]
Comment (idegtiarov) [reply lms/djangoapps/sga_api/views.py]: Got it, that makes sense.

File: common/djangoapps/student/models.py [hunk 50]
```diff
@@ -292,6 +292,8 @@ class Meta(object):
     bio = models.CharField(blank=True, null=True, max_length=3000, db_index=False)
     profile_image_uploaded_at = models.DateTimeField(null=True, blank=True)
 
+    phone = models.CharField(blank=True, max_length=50, db_index=True)
```
Comment (idegtiarov) [inline common/djangoapps/student/models.py]: This field is not required to be indexed.

File: lms/djangoapps/sga_api/README.md [hunk 51]
```diff
@@ -0,0 +1,53 @@
...
+	         "sga_test3@raccoongang.com",
+	         "sga_test20_errr@raccoongang.com"
+	],
+	"courses":[
+		"course-v1:test+test_1+2014_1",
+		"course-v1:test+test_1+2014_5"
+	],
+	"action": "enroll",
+	"email_students": false,
+	"auto_enroll": false
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/README.md]: Did we remove this feature from the API?

File: lms/djangoapps/sga_api/views.py [hunk 52]
```diff
@@ -0,0 +1,317 @@
...
+        "country": (
+            "Country is incorrect or missed: {value}. For checking: Visit https://www.iso.org/obp . "
+            "Click the Country Codes radio option and click the search button."
+        ),
+        "language": "language is incorrect or missed: {value}. It must be:  pt (Portuguese) or en (English)"
+    }
+
+    def post(self, request):
+        """
+        Creates a user using email, login, country and language.
```
Comment (idegtiarov) [inline lms/djangoapps/sga_api/views.py]: I would change s/using/by/
